ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2838,272 @@ public sealed class Bot : IAsyncDisposable {
+			linkElements.AddRange(badgePage.SelectNodes("//div[@class='badges_sheet']/div[contains(@class, 'badge_row') and .//div[@class='badge_info_description']/div[contains(text(), 'Level 5')]]/a[@class='badge_row_overlay']"));
+
+			if (linkElements.Count == 0) {
+				return new HashSet<uint>(0);
+			}
+
+			HashSet<uint> result = new HashSet<uint>(linkElements.Count);
+
+			foreach (string? badgeUri in linkElements.Select(htmlNode => htmlNode.GetAttribute("href"))) {
+				if (string.IsNullOrEmpty(badgeUri)) {
+					ArchiLogger.LogNullError(nameof(badgeUri));
+
+					return null;
+				}
+
+				// URIs to foil badges are the same as for normal badges except they end with "?border=1"
+				string? appIDText = badgeUri.Split('/', StringSplitOptions.RemoveEmptyEntries)[^1].Split('?', StringSplitOptions.RemoveEmptyEntries)[0];
+
+				if (!uint.TryParse(appIDText, out uint appID) || (appID == 0)) {
+					ArchiLogger.LogNullError(nameof(appID));
+
+					return null;
+				}
+
+				result.Add(appID);
+			}
+
+			return result;
+		}
+
+		private bool CompleteSetLootScheduled;
+
+		private async Task SendCompletedSets() {
+			lock (SendCompleteSetsSemaphore) {
+				if (CompleteSetLootScheduled) {
+					return;
+				}
+
+				CompleteSetLootScheduled = true;
+			}
+
+			await SendCompleteSetsSemaphore.WaitAsync().ConfigureAwait(false);
+
+			try {
+				lock (SendCompleteSetsSemaphore) {
+					CompleteSetLootScheduled = false;
+				}
+
+				HashSet<uint>? appIDs = await GetPossiblyCompletedBadgeAppIDs().ConfigureAwait(false);
+
+				if ((appIDs == null) || (appIDs.Count == 0)) {
+					return;
+				}
+
+				HashSet<Steam.Asset> inventory;
+
+				try {
+					inventory = await ArchiWebHandler.GetInventoryAsync(SteamID)
+						.Where(item => item.Tradable && ((item.Type == Steam.Asset.EType.TradingCard) || (item.Type == Steam.Asset.EType.FoilTradingCard)) && appIDs.Contains(item.RealAppID))
+						.ToHashSetAsync()
+						.ConfigureAwait(false);
+				} catch (HttpRequestException e) {
+					ArchiLogger.LogGenericWarningException(e);
+
+					return;
+				} catch (Exception e) {
+					ArchiLogger.LogGenericException(e);
+
+					return;
+				}
+
+				if (inventory.Count == 0) {
+					ArchiLogger.LogGenericWarning(string.Format(Strings.ErrorIsEmpty), nameof(inventory));
+
+					return;
+				}
+
+				const byte minCardsPerBadge = 5;
+				Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), List<uint>> inventorySets = Trading.GetInventorySets(inventory);
+				Dictionary<uint, byte> cardCountPerAppID = await LoadCardsPerSet(appIDs).ConfigureAwait(false);
+				Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), (uint Sets, byte CardsPerSet)> itemsToTakePerInventorySet = inventorySets.Where(kv => kv.Value.Count >= minCardsPerBadge).ToDictionary(kv => kv.Key, kv => (kv.Value.FirstOrDefault(), cardCountPerAppID.TryGetValue(kv.Key.RealAppID, out byte count) ? count : (byte) 0));
```

Comment:
Use `[0]` instead of `FirstOrDefault()`, default value will be `0` and while we're not going to add additional checks against self-logic, we **expect** at least one amount per a key included in non-empty inventory.

Reply:
Once you handle the return value of `0` below, this should also be rewritten to use `cardCountPerAppID[kv.Key.RealAppID]` - we do not expect this to miss some appIDs, and we definitely do not want a freaking value of 0 if it does.